### ISSUE TITLE
Logstash 1.5RC2 fails to stop if redis input is used

### ISSUE THREAD

**Issue by Mister-X-** [elastic/logstash#2871]
When I use redis as input, it fails to stop using service logstash stop (which sends kill -TERM $PID).

When debugging, here is what I get when sending SIGTERM to the process (not started with service):
```
SIGTERM received. Shutting down the pipeline. {:level=>:warn, :file=>"logstash/agent.rb", :line=>"127", :method=>"execute"}
Sending shutdown signal to input thread {:thread=>#<Thread:0x3c8c03ae run>, :level=>:info, :file=>"logstash/pipeline.rb", :line=>"249", :method=>"shutdown"}
```

And it waits forever, doing nothing. I can send more SIGTERM and more of those appear on the screen.

Additionally, I noticed the following a few lines above, when it was logging and I don't know if it's related or not:
```
Starting stale files cleanup cycle {:files=>{"/var/log/logstash/output.log"=>#<IOWriter:0x6fcc6d1e @active=true, @io=#<File:/var/log/logstash/output.log>>}, :level=>:info, :file=>"logstash/outputs/file.rb", :line=>"196", :method=>"close_stale_files"}
```

Here is the first config file:
```
input {
}

filter {
	mutate {
		add_field => {
		"received_at" => "%{@timestamp}"
		"received_from" => "%{host}"
		}
	}
}

output {
	elasticsearch { host => localhost }
	file {
		path => "/var/log/logstash/output.log"
	}
}
```

I know it could be simplified and some stuff in there do not make much sense but I was testing Logstash (and that's the config I had when the bug happened) and you can reproduce the bug every single time with this config.

Here is the second config file:
```
input {
        redis {
                host => "192.168.110.143"
                data_type => 'list'
                key => 'logstash:redis'
                type => 'redis-input'
                password => 'howaboutthisverylongpassword?'
        }
}
```

Another side note is that it seems that the redis plugin only works for the first few minutes. I just changed 2 lines in the configuration file for redis (from the package install on Ubuntu 14.04.2 LTS 64 bit). Let me know if you need the redis config file.

**Comment by suyograo** [comment]
@Mister-X- this maybe related to #2863. We will validate this once we have merge a fix for that

**Comment by Mister-X-** [comment]
Oh sorry, I should have looked at the existing bugs first. Anyway, I managed to reproduce it in another situation, when using stdin input and stdout output (and a few filter), in verbose mode.

**Comment by suyograo** [comment]
no worries. 

we have a fix in https://github.com/elastic/logstash/commit/c07368cc25d977bbda9fe0d7b2c41e0608a414f4 waiting for reviews to merge

**Comment by colinsurprenant** [comment]
2 different issues here:
- **stdin** input : this is a long standing problem for which we do not have a fix yet. In a nutshell, in Java blocking read is non interruptible, see my comment to that effect in #1769 

- **redis** input : we recently fixed a bunch of issues related to the shutdown sequence in both #2863 and #2928 and also some fixes in the redis input in logstash-plugins/logstash-input-redis#5 - I tested locally and sending `kill -TERM` to logstash with your example config works as intended!

We have a RC3 release coming out soon that will include all these fixes.
